File: aocpy/task4.py
data = open("task4.txt", "r").readlines()

nums = [int(v) for v in data[0].split(",")]

boards = [list(map(lambda x: [int(v) for v in x.split()], data[x:x+5])) for x in range(2, len(data), 6)]

def is_bingo(board, marked):
    for row in board:
        if all(v in marked for v in row):
            return True
        
    for col in range(len(board[0])):
        if all(b[i][col] in marked for i in range(len(board))):
            return True
            
    return False
    
def product_unmarked(board, marked):
    acc = 0
    for row in board:
        for v in row:
            if v not in marked:
                acc += v
                
    return acc
    
marked = []
found = False
for n in nums:
    marked.append(n)
    print(marked)
    for b in boards:
        if is_bingo(b, marked):
            print(product_unmarked(b, marked) * marked[-1])
            found = True
            break
    if found:
        break
        
marked = []
won = []
for n in nums:
    marked.append(n)
    print(marked)
    all_won = True
    for b in boards:
        if b in won:
            continue
        if is_bingo(b, marked):
            won.append(b)
        all_won = False
    if all_won:
        print(product_unmarked(won[-1], marked[:-1]), marked[-2])
        print(product_unmarked(won[-1], marked[:-1]) * marked[-2])
        break
            
            
            

File: src/task4.rs
use std::mem::MaybeUninit;

use crate::util::get_task;

type Bingarr = [[usize; 5]; 5];

fn is_bingo(bingo: &Bingarr, marked: &[usize]) -> bool {
    bingo.iter().any(|r| r.iter().all(|v| marked.contains(v)))
        || (0..bingo.len()).any(|i| bingo.iter().all(|r| marked.contains(&r[i])))
}

fn unmarked_sum(bingo: &Bingarr, marked: &[usize]) -> usize {
    bingo
        .iter()
        .flat_map(|r| r.iter())
        .filter(|v| !marked.contains(v))
        .sum()
}

pub fn task4() {
    // io
    let input = get_task(4);
    let (l1, rest) = input.split_once("\n\n").unwrap();
    let nums: Vec<usize> = l1
        .split(',')
        .map(|v| v.parse().unwrap())
        .collect();
        
    let bingos: Vec<[[usize; 5]; 5]> = rest.split("\n\n")
        .map(|chnk| {
            chnk.lines()
                .map(|l| l.split_whitespace().map(|v| v.parse().unwrap()))
        })
        .map(|bingo| {
            // Create an uninitialized array of `MaybeUninit`. The `assume_init` is
            // safe because the type we are claiming to have initialized here is a
            // bunch of `MaybeUninit`s, which do not require initialization.
            let mut out: [[MaybeUninit<usize>; 5]; 5] = unsafe {
                MaybeUninit::uninit().assume_init()
            };
            for (row, line) in bingo.enumerate() {
                for (col, val) in line.enumerate() {
                    out[row][col].write(val);
                }
            }
            // all initialized, so we can safely convert
            unsafe { std::mem::transmute(out) }
        })
        .collect();

    // task 1

    'outer1: for i in 1..nums.len() {
        let marked = &nums[..i];
        for b in &bingos {
            if is_bingo(b, marked) {
                let res = unmarked_sum(b, marked) * marked.last().unwrap();
                println!("part 1: {}", res);
                break 'outer1;
            }
        }
    }

    // task 2

    let mut won_bingos = vec![false; bingos.len()];
    let mut count = 0;
    'outer2: for i in 1..nums.len() {
        let marked = &nums[..i];
        for (i, b) in bingos.iter().enumerate() {
            if won_bingos[i] {
                continue;
            }
            if is_bingo(b, marked) {
                won_bingos[i] = true;
                count += 1;
            }
            if count == bingos.len() {
                let res = unmarked_sum(b, marked) * marked.last().unwrap();
                println!("part 2: {}", res);
                break 'outer2;
            }
        }
    }
}
